librad/src/id/entity/data.rs

```diff
@@ -0,0 +1,209 @@
+where
+    S: Serializer,
+{
+    let ordered: BTreeSet<String> = value.iter().cloned().collect();
+    ordered.serialize(serializer)
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, Default)]
+pub struct EntityData<T> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub revision: Option<u64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub hash: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parent_hash: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub signatures: Option<HashMap<String, EntitySignatureData>>,
+
+    #[serde(
+        skip_serializing_if = "HashSet::is_empty",
+        serialize_with = "ordered_set",
+        default
+    )]
+    pub keys: HashSet<String>,
+    #[serde(
+        skip_serializing_if = "HashSet::is_empty",
+        serialize_with = "ordered_set",
+        default
+    )]
+    pub certifiers: HashSet<String>,
+
+    pub info: T,
+}
+
+impl<T> EntityData<T>
+where
+    T: Serialize + DeserializeOwned + Clone + Default,
+{
+    pub fn to_json_writer<W>(&self, writer: W) -> Result<(), Error>
+    where
+        W: std::io::Write,
+    {
+        serde_json::to_writer(writer, self).map_err(Error::SerializationFailed)?;
+        Ok(())
+    }
+    pub fn to_json_string(&self) -> Result<String, Error> {
+        Ok(serde_json::to_string(self).map_err(Error::SerializationFailed)?)
+    }
+
+    pub fn from_json_reader<R>(r: R) -> Result<Self, Error>
+    where
+        R: std::io::Read,
+    {
+        Ok(serde_json::from_reader(r).map_err(Error::SerializationFailed)?)
+    }
+    pub fn from_json_str(s: &str) -> Result<Self, Error> {
+        Ok(serde_json::from_str(s).map_err(Error::SerializationFailed)?)
+    }
+
+    pub fn canonical_data(&self) -> Result<Vec<u8>, Error> {
+        let mut cleaned = EntityData::<T>::default();
+        cleaned.name = self.name.to_owned();
+        cleaned.revision = self.revision.to_owned();
+        cleaned.hash = self.hash.to_owned();
+        cleaned.keys = self.keys.to_owned();
+        cleaned.certifiers = self.certifiers.to_owned();
+        cleaned.info = self.info.to_owned();
+
+        let mut buffer: Vec<u8> = vec![];
+        let mut ser =
+            serde_json::Serializer::with_formatter(&mut buffer, CanonicalFormatter::new());
+        cleaned
+            .serialize(&mut ser)
+            .map_err(Error::SerializationFailed)?;
+        Ok(buffer)
+    }
+
+    pub fn compute_hash(&self) -> Result<Multihash, Error> {
+        Ok(Sha2_256::digest(&self.canonical_data()?))
+    }
+
+    pub fn new(name: String, revision: u64) -> Self {
+        let mut result = Self::default();
+        result.name = Some(name);
+        result.revision = Some(revision);
+        result
+    }
+
+    pub fn set_name(mut self, name: String) -> Self {
```

Comment:
This is meant to be used as a "builder":
```
User::default()
  .set_something(...)
  .set_something_else(...)
  .build()?
```

So all the "setters" need to consume the data and return a "new one", of course they actually just move a mutated one.

Properties are public because this is just a temporary data structure with no invariant at  all, it is meant to be used in the most convenient way when building a new entity.

Reply:
Ah, my bad! In my head, I was thinking of `&mut` instead of just `mut`. This makes sense now :ok_hand: Although I wonder if it's more "idiomatic" to use a reference and mutate that. Calling resident Rust aficionado @cloudhead :eyes:

Reply:
I still think that having the fields public _and_ having the setters doesn't make much sense. I'm thinking of it in the sense of how someone should use this API. Should they set the fields manually or use the functions? Pick one, so that they don't have to guess and there isn't a disagreement on which is the right way :)

Reply:
I don't have a strong opinion on this, but I suspect that both are needed here.

The "mutator" methods consume the input and produce a new one, so they can *only* be used in a "chained invocation" way, where the last invocation is the builder or anyway something that consumes the value.

On the other hand public properties allow a more traditional "make a value, do something with it, then so something else" coding style.

I don't know exactly how this API will be used so I provided both so that the API user can choose the style.

In general public properties would be enough (once you have them you can do whatever you want) but I see that the "chained mutations on a builder" is a common pattern and leads to more readable code so I provided that as well.

Reply:
Maybe it's at this point where I need to understand why this data structure exists in the first place :)

Reply:
Re: builders: both forms are okay, the only advantage of borrowing is that you can *also* use it this way:
```rust
let builder = Builder::new();
builder.set("foo", "bar");
builder.set("quk", "bar");
builder.build()
```
Otherwise, you're forced to use a one-liner approach or repeat the `let` binding a bunch of times.

Reply:
Oh, I see... I am allowing that pattern by having public properties.
Without public properties I would need to borrow instead of move in all setters, *and* provide getters.

In this case, where all the invariants are checked at the end anyway, I feel that having public properties is more flexible and does not have downsides.

But I don't have strong opinions, if asked I can make setters borrow (take and retuen a `&ref mut`) and add getters.
The `build` method would still move and consume the builder instance (closing the build).